### ISSUE TITLE
Linking error for iOS with SQLitePCLRaw 1.0 - MTOUCH error MT5214

### ISSUE THREAD

**Issue by nathanielcook** [ericsink/SQLitePCL.raw#99]
Attempting to upgrade from v0.9.3 to 1.0.0. Android is working fine, but I'm getting a ton of errors like the following at compile time when I build iOS (unified):

```
MTOUCH: error MT5214: Native linking failed, undefined symbol: _sqlite3_aggregate_context. This symbol was referenced by the managed member SQLitePCL.SQLite3Provider_internal/NativeMethods.sqlite3_aggregate_context. Please verify that all the necessary frameworks have been referenced and native libraries linked.
etc
MTOUCH: error MT5214: Native linking failed, undefined symbol: _sqlite3_table_column_metadata. This symbol was referenced by the managed member SQLitePCL.SQLite3Provider_internal/NativeMethods.sqlite3_table_column_metadata. Please verify that all the necessary frameworks have been referenced and native libraries linked.
MTOUCH: error MT5214: Native linking failed, undefined symbol: _sqlite3_threadsafe. This symbol was referenced by the managed member SQLitePCL.SQLite3Provider_internal/NativeMethods.sqlite3_threadsafe. Please verify that all the necessary frameworks have been referenced and native libraries linked.
MTOUCH: error MT5214: Native linking failed, undefined symbol: _sqlite3_total_changes. This symbol was referenced by the managed member 
etc
```
Looks like there are about 130 of them. It's as if the actual SQLite binary has not been included at all. I'm trying to bundle a version of SQLite on iOS rather than use the built-in SQLite. It was working fine with 0.9.3.

Here are the Nuget packages I'm using:
```
  <package id="SQLitePCLRaw.core" version="1.0.0" targetFramework="xamarinios10" />
  <package id="SQLitePCLRaw.lib.e_sqlite3.ios_unified.static" version="1.0.0" targetFramework="xamarinios10" />
  <package id="SQLitePCLRaw.provider.internal.ios_unified" version="1.0.0" targetFramework="xamarinios10" />
```

I tried adding package `<package id="SQLitePCLRaw.bundle_e_sqlite3" version="1.0.0" targetFramework="xamarinios10" />` and using `SQLitePCL.Batteries.Init()` instead of `SQLitePCL.raw.SetProvider(new SQLitePCL.SQLite3Provider_internal())` but that didn't help. Same results. (It's probably worth mentioning that if I don't set any provider or call Batteries.Init then the project compiles successfully, but then as expected I get a runtime error saying I didn't set any provider. Also bundle_green works fine.)

So I feel like I might be doing something wrong, so let me ask a couple of questions that might clear up my understanding:

- Why is the provider for iOS included with bundle_e_sqlite3 called SQLite3Provider_internal instead of SQLite3Provider_e_sqlite3? (I don't see anything in the documentation explaining what internal means.) 
- Is there an e_sqlite3 provider for iOS? I can't find one on Nuget. 
- Out of curiosity why does the lib package for ios have `.static` in the name?

**Comment by ericsink** [comment]
I'm guessing you are actually not doing anything wrong.  This is very likely a bug that slipped through.  But I'll test and verify that.

For now, responding to your questions:

iOS is a special case.  Bundling a SQLite library on iOS is done with static linking, instead of the dynamic linking that is used everywhere else.

At the pinvoke level, the way this mechanism works is to DllImport("__Internal"), which tells Xamarin.iOS to look for the symbol as it came from a statically linked library.  

If instead it said DllImport("e_sqlite3"), then it would look for the symbol in a dynamic library by that name.  Which, in theory, can work.  In some cases.  After iOS 8.  But I haven't sorted out all the somewhat-newfangled support for dynamic libs on iOS.

So, no, there currently is no e_sqlite3 provider for iOS, at least not one which does dynamic loading.  On iOS, e_sqlite3 is provided only as a static lib (and that's why the word static is in the name of that package) for now, which means that provider.internal is needed.

In other words, for the purpose of bundle_e_sqlite3, provider.internal (with that lib) *is* the e_sqlite3 provider.

And it looks like it might be broken.

**Comment by ericsink** [comment]
Confirmed. I can reproduce this. Apologies. Fix pending.

**Comment by ericsink** [comment]
I've confirmed that the "attempted fix" in dc98825 was successful.  I will mark this closed when I have pushed 1.0.1 to nuget, which should happen a few days from now unless somebody tells me it's more urgent than that.

**Comment by nathanielcook** [comment]
Thanks for looking into this so quickly! We can wait a few days.

**Comment by ericsink** [comment]
Fixed in 1.0.1, which has been pushed to nuget.org. I'll leave this open for now so you can close it after verifying that 1.0.1 works as expected for you.

**Comment by ps30** [comment]
Hey Eric - Not seeing any new packages on NuGet (using Visual Studio 2015 to check update on packages).  Here is what I see when checking for updates - all still v0.9.4.-pre1:
![image](https://cloud.githubusercontent.com/assets/12287002/18679381/42729d28-7f14-11e6-933e-3bac12166eae.png)

**Comment by ericsink** [comment]
The package IDs have changed.  And I have not yet marked all the old ones with "deprecated" (or some other way of referring people to the new ones).  But I need to do that very soon.

All the pkgs now have an id that begins with "SQLitePCLRaw."

So in your screen shot, change to SQLitePCLRaw.bundle_green

**Comment by ps30** [comment]
OK, and it looks like before we can do that Frank Krueger (**praeclarum/sqlite-net**) needs to update to support the new names in his package.  Have you pinged him re name changes?
![image](https://cloud.githubusercontent.com/assets/12287002/18681041/191b83c0-7f1b-11e6-9be0-16c51840aa7a.png)

**Comment by ericsink** [comment]
praeclarum/sqlite-net#449

**Comment by nathanielcook** [comment]
Interesting. It works for me now but only if I reference `bundle_e_sqlite3` and call `Batteries.Init();`. If I call `SQLitePCL.raw.SetProvider(new SQLitePCL.SQLite3Provider_internal());` then I get the same compiler error as before. Very strange.

**Comment by ericsink** [comment]
Yeah, that won't work.  :-)

The problem is the Xamarin linker doesn't think you need the assembly, so it removes it.

So for this case, in addition to raw.SetProvider(), you have to call:

    SQLitePCL.lib.embedded.Init();

which is a do-nothing method in the assembly where the native lib is embedded.  This tricks the Xamarin linker.  (FWIW, I've seen this trick used in a lot of other places.  MvvmCross uses it quite a bit.)

bundle_e_sqlite3 does this call for you.  That's why it works with the bundle.

**Comment by nathanielcook** [comment]
Ah, I saw that in batteries.cs file but didn't think to try it. Great will go ahead and close this. I recommend that we include some of the iOS information here (about the meaning of static/internal, and that SetProvider alone won't work) in the readme. Thank you
